src/main/kotlin/eu/europa/ec/eudi/pidissuer/adapter/out/jose/VerifyKeyAttestation.kt

```
@@ -0,0 +1,214 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package eu.europa.ec.eudi.pidissuer.adapter.out.jose
+
+import arrow.core.*
+import com.nimbusds.jose.JOSEObjectType
+import com.nimbusds.jose.JWSAlgorithm
+import com.nimbusds.jose.crypto.ECDSASigner
+import com.nimbusds.jose.crypto.Ed25519Signer
+import com.nimbusds.jose.crypto.RSASSASigner
+import com.nimbusds.jose.jwk.*
+import com.nimbusds.jose.proc.DefaultJOSEObjectTypeVerifier
+import com.nimbusds.jose.proc.SecurityContext
+import com.nimbusds.jose.proc.SingleKeyJWSKeySelector
+import com.nimbusds.jose.util.Base64
+import com.nimbusds.jose.util.X509CertChainUtils
+import com.nimbusds.jose.util.X509CertUtils
+import com.nimbusds.jwt.JWTClaimsSet
+import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier
+import com.nimbusds.jwt.proc.DefaultJWTProcessor
+import eu.europa.ec.eudi.pidissuer.adapter.out.util.getOrThrow
+import eu.europa.ec.eudi.pidissuer.domain.KeyAttestation
+import eu.europa.ec.eudi.pidissuer.domain.KeyAttestationJWT
+import eu.europa.ec.eudi.pidissuer.port.out.credential.VerifyCNonce
+import java.net.URI
+import java.security.KeyStore
+import java.security.cert.*
+import java.time.Instant
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.DurationUnit
+
+internal val SkipRevocation: PKIXParameters.() -> Unit = { isRevocationEnabled = false }
+
+internal class VerifyKeyAttestation(
+    private val trustAnchors: KeyStore? = null,
+    private val verifyAttestedKey: (JWK) -> JWK?,
```

The abstraction is problematic

- it is lossy. Totally discards the error channel.
- it is not suspend function
- Most importantly, it doesn't take into account the attestation requirements as defined by the issuer


Please consider introducing a function interface 

```kotlin
fun interface VerifyAttestedKey {
    suspend fun verify(key: JWK, requirement: KeyAttestation.Required): Either<SomeError, JWK>
    suspend fun verify(keys: NomEmptyList<JWK>, requirement: KeyAttestation.Required): Either<SomeError, JWK> = TODO ("Implement this using arrow, accumulating errors")
}


```

---

Check also my comment about `cNonce`

Please do not remove it. Just  define the interface